pkg/trustedresources/verifier/verifier.go

```diff
@@ -0,0 +1,192 @@
+			}
+			verifiers = append(verifiers, v)
+		} else if a.Key.SecretRef != nil {
+			v, err := fromSecret(ctx, fmt.Sprintf("%s%s/%s", keyReference, a.Key.SecretRef.Namespace, a.Key.SecretRef.Name), algorithm, k8s)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get verifier from secret: %w", err)
+			}
+			verifiers = append(verifiers, v)
+		}
+	}
+	if len(verifiers) == 0 {
+		return verifiers, errorEmptyPublicKeys
+	}
+	return verifiers, nil
+
+}
+
+// fromKeyRef parses the given keyRef, loads the key and returns an appropriate
+// verifier using the provided hash algorithm
+func fromKeyRef(ctx context.Context, keyRef string, hashAlgorithm crypto.Hash, k8s kubernetes.Interface) (verifier signature.Verifier, err error) {
+	var raw []byte
+	if strings.HasPrefix(keyRef, keyReference) {
+		v, err := fromSecret(ctx, keyRef, hashAlgorithm, k8s)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get verifier from secret: %w", err)
+		}
+		return v, nil
+	}
+	raw, err = os.ReadFile(filepath.Clean(keyRef))
+	if err != nil {
+		return nil, fmt.Errorf("%w: %v", ErrorFailedLoadKeyFile, err)
+	}
+	v, err := fromData(raw, hashAlgorithm)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get verifier from data: %w", err)
+	}
+	return v, nil
+}
+
+// fromSecret fetches the public key from SecretRef and returns the verifier
+// hashAlgorithm is provided to determine the hash algorithm of the key
+func fromSecret(ctx context.Context, secretRef string, hashAlgorithm crypto.Hash, k8s kubernetes.Interface) (signature.Verifier, error) {
+	if strings.HasPrefix(secretRef, keyReference) {
+		s, err := keyPairSecret(ctx, secretRef, k8s)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get secret: %w", err)
+		}
+		// only 1 public key should be in the secret
+		if len(s.Data) == 0 {
+			return nil, fmt.Errorf("secret %q contains no data %w", secretRef, errorEmptySecretData)
+		}
+		if len(s.Data) > 1 {
+			return nil, fmt.Errorf("secret %q contains multiple data entries, only one is supported. %w", secretRef, errorMultipleSecretData)
+		}
+		for _, raw := range s.Data {
+			v, err := fromData(raw, hashAlgorithm)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get verifier from secret data: %w", err)
+			}
+			return v, nil
+		}
+	}
+	return nil, fmt.Errorf("%w: secretRef %v is invalid", errorK8sSpecificationInvalid, secretRef)
+}
+
+// fromData fetches the public key from raw data and returns the verifier
+func fromData(raw []byte, hashAlgorithm crypto.Hash) (signature.Verifier, error) {
+	pubKey, err := cryptoutils.UnmarshalPEMToPublicKey(raw)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %v", ErrorDecodeKey, err)
+	}
+	v, err := signature.LoadVerifier(pubKey, hashAlgorithm)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %v", errorLoadVerifier, err)
+	}
+	return v, nil
+}
+
+// keyPairSecret fetches the secret from a k8sRef
+func keyPairSecret(ctx context.Context, k8sRef string, k8s kubernetes.Interface) (*v1.Secret, error) {
```

Comment:
can you give this a name which includes a verb to indicate what it is doing? e.g. something like `getKeyPairSecret` (looking at the current name I wouldn't have guessed that it is actually making a call to the API server - and p.s. do we want to use a lister here?)

Reply:
Thanks for the suggestion! `lister` is better but we will need to pass it from reconciler. How about adding a todo here? There may be better ways. I can create an issue to list possible solutions

Reply:
since you are already passing the k8s client all the way through, i think it shouldnt be too much worse to pass the lister through - specfically a Get function from the lister, which would actually be easier to mock in your unit tests (you  just need to make a function with the same signature)

@Yongxuanzhang pointed out difficulty around using the namespace to get the correct lister function, i think https://github.com/tektoncd/pipeline/blob/f732985ec30817703e6eeaa6914cc77a2b8adf3c/pkg/reconciler/pipelinerun/pipelinerun.go#L331 is an example of working around that (wrapping the function such that you have the namespace)